src/OpenTelemetry/Logs/Pool/LogRecordSharedPool.cs

```diff
@@ -0,0 +1,172 @@
+// <copyright file="LogRecordSharedPool.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+#nullable enable
+
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+
+namespace OpenTelemetry.Logs
+{
+    internal sealed class LogRecordSharedPool : ILogRecordPool
+    {
+        internal const int DefaultMaxPoolSize = 2048;
+        internal const int DefaultMaxNumberOfAttributes = 64;
+        internal const int DefaultMaxNumberOfScopes = 16;
+
+        internal static LogRecordSharedPool Current = new(DefaultMaxPoolSize);
+
+        internal readonly int Capacity;
+        private readonly LogRecord?[] pool;
+        private long rentIndex;
+        private long returnIndex;
+
+        internal LogRecordSharedPool(int capacity)
+        {
+            this.Capacity = capacity;
+            this.pool = new LogRecord?[capacity];
+        }
+
+        internal int Count => (int)(Volatile.Read(ref this.returnIndex) - Volatile.Read(ref this.rentIndex));
+
+        public LogRecord Rent()
+        {
+            while (true)
+            {
+                var rentSnapshot = Volatile.Read(ref this.rentIndex);
+                var returnSnapshot = Volatile.Read(ref this.returnIndex);
+
+                if (rentSnapshot >= returnSnapshot)
+                {
+                    break; // buffer is empty
+                }
+
+                if (Interlocked.CompareExchange(ref this.rentIndex, rentSnapshot + 1, rentSnapshot) == rentSnapshot)
+                {
+                    var logRecord = Interlocked.Exchange(ref this.pool[rentSnapshot % this.Capacity], null);
+                    if (logRecord == null && !this.TryRentCoreRare(rentSnapshot, out logRecord))
```

Comment:
I think this can be simplified by using `CompareExchange` instead of `Exchange` for retrieving a logRecord object from the pool:

```diff
+            LogRecord? logRecord = null;
            while (true)
            {
                var rentSnapshot = Volatile.Read(ref this.rentIndex);
                var returnSnapshot = Volatile.Read(ref this.returnIndex);
+               var recordToRent = this.pool[rentSnapshot % this.Capacity];

                if (rentSnapshot >= returnSnapshot)
                {
                    break; // buffer is empty
                }

                if (Interlocked.CompareExchange(ref this.rentIndex, rentSnapshot + 1, rentSnapshot) == rentSnapshot)
                {
-                    var logRecord = Interlocked.Exchange(ref this.pool[rentSnapshot % this.Capacity], null);
-                    if (logRecord == null && !this.TryRentCoreRare(rentSnapshot, out logRecord))
-                    {
-                        continue;
-                    }
+                    if (Interlocked.CompareExchange(ref this.pool[rentSnapshot % this.Capacity], null, recordToRent) == recordToRent)
+                    {
+                        logRecord = recordToRent;
+                    }

+                    if (logRecord == null)
+                    {
+                        continue;
+                    }

                    logRecord!.ResetReferenceCount();

                    return logRecord;
                }
            }
```

Reply:
Sure.

> That could lead to a higher rate of misses under load which would reduce throughput. But harder to test that.

I think we could use some Stress Tests to measure that later.